token/services/tokens/tokens.go

```diff
@@ -0,0 +1,193 @@
+	IsMine(tms *token.ManagementService, tok *token2.Token) ([]string, bool)
+	// AmIAnAuditor return true if the passed TMS contains an auditor wallet for any of the auditor identities
+	// defined in the public parameters of the passed TMS.
+	AmIAnAuditor(tms *token.ManagementService) bool
+}
+
+type Issued interface {
+	// Issued returns true if the passed issuer issued the passed token
+	Issued(tms *token.ManagementService, issuer view.Identity, tok *token2.Token) bool
+}
+
+type GetTMSProviderFunc = func() *token.ManagementServiceProvider
+
+// Transaction models a token transaction
+type Transaction interface {
+	ID() string
+	Network() string
+	Channel() string
+	Namespace() string
+	Request() *token.Request
+}
+
+// Tokens is the interface for the owner service
+type Tokens struct {
+	TMSProvider TMSProvider
+	Ownership   Authorization
+	Issued      Issued
+	TokenStore  TokenStore
+}
+
+func (t *Tokens) AppendTransaction(tx Transaction) error {
+	txID := tx.ID()
+	tms, err := t.TMSProvider.GetManagementService(
+		token.WithNetwork(tx.Network()),
+		token.WithChannel(tx.Channel()),
+		token.WithNamespace(tx.Namespace()),
+	)
+	if err != nil {
+		return errors.WithMessagef(err, "failed getting token management service [%s:%s:%s]", tx.Network(), tx.Channel(), tx.Namespace())
+	}
+	if logger.IsEnabledFor(zapcore.DebugLevel) {
+		logger.Debugf("transaction [%s on (%s)] is known, extract tokens", txID, tms.ID())
+	}
+	request := tx.Request()
+	if request.Metadata == nil {
+		logger.Debugf("transaction [%s], no metadata found, skip it", txID)
+		return nil
+	}
+	metadata, err := request.GetMetadata()
+	if err != nil {
+		logger.Debugf("transaction [%s], failed to get metadata [%s]", txID, err)
+		return err
+	}
+
+	precision := tms.PublicParametersManager().PublicParameters().Precision()
+	if tms.PublicParametersManager().PublicParameters().GraphHiding() {
+		ids := metadata.SpentTokenID()
+		if logger.IsEnabledFor(zapcore.DebugLevel) {
+			logger.Debugf("transaction [%s] with graph hiding, delete inputs [%v]", txID, ids)
+		}
+		for _, id := range ids {
+			if err := t.TokenStore.DeleteToken(id.TxId, id.Index, txID); err != nil {
+				return err
+			}
+		}
+	}
+
+	auditorFlag := t.Ownership.AmIAnAuditor(tms)
+	if auditorFlag {
+		if logger.IsEnabledFor(zapcore.DebugLevel) {
+			logger.Debugf("transaction [%s], I must be the auditor", txID)
+		}
+	}
+
+	is, os, err := request.InputsAndOutputs()
+	if err != nil {
+		return errors.WithMessagef(err, "failed to get request's outputs")
+	}
+
+	// parse the inputs
+	for _, input := range is.Inputs() {
+		if input.Id == nil {
+			logger.Infof("transaction [%s] found an input that is not mine, skip it", tx.ID())
+			continue
+		}
+		if logger.IsEnabledFor(zapcore.DebugLevel) {
+			logger.Debugf("transaction [%s] delete input [%s]", tx.ID(), input.Id)
+		}
+		if err := t.TokenStore.DeleteToken(input.Id.TxId, input.Id.Index, tx.ID()); err != nil {
```

Comment:
maybe for better rollback it would make sense to collect all tokens to be deleted and do them in one transaction at the end?

Reply:
yes, I mentioned this in the description of the PR as a point to address in a subsequent PR 👍